### ISSUE TITLE
Text Predictions Tutorials Has  Incorrect Data

### ISSUE THREAD

**Issue by kn0w1dg3** [tensorflow/tensorflow#59184]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

No

### Source

source

### Tensorflow Version

2.8

### Custom Code

No

### Current Behaviour?

```shell
A bug happened! On the Tutorial page, the predictions for generating text is as follows.

To get actual predictions from the model you need to sample from the output distribution, to get actual character indices. This distribution is defined by the logits over the character vocabulary.

Note: It is important to sample from this distribution as taking the argmax of the distribution can easily get the model stuck in a loop.
Try it for the first example in the batch:


sampled_indices = tf.random.categorical(example_batch_predictions[0], num_samples=1)
sampled_indices = tf.squeeze(sampled_indices, axis=-1).numpy()
This gives us, at each timestep, a prediction of the next character index:


sampled_indices

array([49, 56, 39, 28,  8, 35, 34,  0, 11, 12, 34, 36,  4, 34, 50, 12, 12,
       54, 28, 46, 24, 41, 29, 16, 11, 59,  5, 62, 15, 33,  8, 61,  3, 30,
       10, 37, 57, 48, 41, 61, 11, 55, 43, 45, 57, 51,  6, 36,  8, 18, 32,
       30, 49,  2, 35, 47, 25, 51, 43, 49, 11, 50, 64, 37, 34, 29, 43,  4,
       51, 19,  9, 12, 18, 48, 31,  7, 35, 30, 38,  1, 35, 49, 22, 65, 48,
       13, 26, 12, 56, 25, 30, 41, 28, 27,  1, 62, 41, 34,  4, 29])
Decode these to see the text predicted by this untrained model:


print("Input:\n", text_from_ids(input_example_batch[0]).numpy())
print()
print("Next Char Predictions:\n", text_from_ids(sampled_indices).numpy())

Input:
 b"! I come, I come!\nWho knocks so hard? whence come you? what's your will?\n\nNurse:\n\nFRIAR LAURENCE:\nWe"
Next Char Predictions:
 b"jqZO-VU[UNK]:;UW\\(Uk;;oOgKbPC:t&wBT-v!Q3Xribv:pdfrl'W-ESQj VhLldj:kyXUPd\\)lF.;EiR,VQY\nVjIzi?M;qLQbON\nwbU$P"


This is within training, and even me who is new, knows this is wrong.
```

### Standalone code to reproduce the issue

```shell
Try the model
Now run the model to see that it behaves as expected.

First check the shape of the output:


for input_example_batch, target_example_batch in dataset.take(1):
    example_batch_predictions = model(input_example_batch)
    print(example_batch_predictions.shape, "# (batch_size, sequence_length, vocab_size)")

(64, 100, 66) # (batch_size, sequence_length, vocab_size)
In the above example the sequence length of the input is 100 but the model can be run on inputs of any length:


model.summary()

Model: "my_model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 embedding (Embedding)       multiple                  16896     
                                                                 
 gru (GRU)                   multiple                  3938304   
                                                                 
 dense (Dense)               multiple                  67650     
                                                                 
=================================================================
Total params: 4,022,850
Trainable params: 4,022,850
Non-trainable params: 0
_________________________________________________________________
To get actual predictions from the model you need to sample from the output distribution, to get actual character indices. This distribution is defined by the logits over the character vocabulary.

Note: It is important to sample from this distribution as taking the argmax of the distribution can easily get the model stuck in a loop.
Try it for the first example in the batch:


sampled_indices = tf.random.categorical(example_batch_predictions[0], num_samples=1)
sampled_indices = tf.squeeze(sampled_indices, axis=-1).numpy()
This gives us, at each timestep, a prediction of the next character index:


sampled_indices

array([49, 56, 39, 28,  8, 35, 34,  0, 11, 12, 34, 36,  4, 34, 50, 12, 12,
       54, 28, 46, 24, 41, 29, 16, 11, 59,  5, 62, 15, 33,  8, 61,  3, 30,
       10, 37, 57, 48, 41, 61, 11, 55, 43, 45, 57, 51,  6, 36,  8, 18, 32,
       30, 49,  2, 35, 47, 25, 51, 43, 49, 11, 50, 64, 37, 34, 29, 43,  4,
       51, 19,  9, 12, 18, 48, 31,  7, 35, 30, 38,  1, 35, 49, 22, 65, 48,
       13, 26, 12, 56, 25, 30, 41, 28, 27,  1, 62, 41, 34,  4, 29])
Decode these to see the text predicted by this untrained model:


print("Input:\n", text_from_ids(input_example_batch[0]).numpy())
print()
print("Next Char Predictions:\n", text_from_ids(sampled_indices).numpy())

Input:
 b"! I come, I come!\nWho knocks so hard? whence come you? what's your will?\n\nNurse:\n\nFRIAR LAURENCE:\nWe"

Next Char Predictions:
 b"jqZO-VU[UNK]:;UW\\(Uk;;oOgKbPC:t&wBT-v!Q3Xribv:pdfrl'W-ESQj VhLldj:kyXUPd\\)lF.;EiR,VQY\nVjIzi?M;qLQbON\nwbU$P"
```

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@kn0w1dg3,
Could you please share the mentioned Text prediction link where the requested change is to be made which helps to debug the issue. Thank you!

**Comment by kn0w1dg3** [comment]
This came from tensorflows tutorial area for text generation with an RNN.This is the try the model section.Try the modelNow run the model to see that it behaves as expected.First check the shape of the output:for input_example_batch, target_example_batch in dataset.take(1):    example_batch_predictions = model(input_example_batch)    print(example_batch_predictions.shape, "# (batch_size, sequence_length, vocab_size)")(64, 100, 66) # (batch_size, sequence_length, vocab_size)In the above example the sequence length of the input is 100 but the model can be run on inputs of any length:model.summary()Model: "my_model"_________________________________________________________________ Layer (type)                Output Shape              Param #   ================================================================= embedding (Embedding)       multiple                  16896                                                                       gru (GRU)                   multiple                  3938304                                                                     dense (Dense)               multiple                  67650                                                                      =================================================================Total params: 4,022,850Trainable params: 4,022,850Non-trainable params: 0_________________________________________________________________To get actual predictions from the model you need to sample from the output distribution, to get actual character indices. This distribution is defined by the logits over the character vocabulary.Note: It is important to sample from this distribution as taking the argmax of the distribution can easily get the model stuck in a loop.Try it for the first example in the batch:sampled_indices = tf.random.categorical(example_batch_predictions[0], num_samples=1)sampled_indices = tf.squeeze(sampled_indices, axis=-1).numpy()This gives us, at each timestep, a prediction of the next character index:sampled_indicesarray([49, 56, 39, 28,  8, 35, 34,  0, 11, 12, 34, 36,  4, 34, 50, 12, 12,       54, 28, 46, 24, 41, 29, 16, 11, 59,  5, 62, 15, 33,  8, 61,  3, 30,       10, 37, 57, 48, 41, 61, 11, 55, 43, 45, 57, 51,  6, 36,  8, 18, 32,       30, 49,  2, 35, 47, 25, 51, 43, 49, 11, 50, 64, 37, 34, 29, 43,  4,       51, 19,  9, 12, 18, 48, 31,  7, 35, 30, 38,  1, 35, 49, 22, 65, 48,       13, 26, 12, 56, 25, 30, 41, 28, 27,  1, 62, 41, 34,  4, 29])Decode these to see the text predicted by this untrained model:print("Input:\n", text_from_ids(input_example_batch[0]).numpy())print()print("Next Char Predictions:\n", text_from_ids(sampled_indices).numpy())Input: b"! I come, I come!\nWho knocks so hard? whence come you? what's your will?\n\nNurse:\n\nFRIAR LAURENCE:\nWe" Next Char Predictions: b"jqZO-VU[UNK]:;UW\\(Uk;;oOgKbPC:t&wBT-v!Q3Xribv:pdfrl'W-ESQj VhLldj:kyXUPd\\)lF.;EiR,VQY\nVjIzi?M;qLQbON\nwbU$P" From: tilakrayalSent: Tuesday, January 10, 2023 1:34 AMTo: tensorflow/tensorflowCc: Eric Boudreau; MentionSubject: Re: [tensorflow/tensorflow] Text Predictions Tutorials Has Incorrect Data (Issue ***@***.***,Could you please share the mentioned Text prediction link where the requested change is to be made which helps to debug the issue. Thank you!—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

**Comment by Rishikesh2338** [comment]
Could you please assign me this work to fix the work

**Comment by tilakrayal** [comment]
@kn0w1dg3,
I was not able to find the official doc link for the Text predictions tutorial. If you are mentioning this text generation official doc link, I was able to execute the code and also the output was also as expected. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/3cc7a56dbab63a49a9e2e6d68386fae7/text_generation.ipynb). Thank you!
